### PR TITLE
Proximal Policy Optimization

### LINKED ISSUE #684: Proximal Policy Optimization
https://blog.openai.com/openai-baselines-ppo/

OpenAI says PPO has become their default RL algorithm. Should we get a PPO implementation going in TensorGraph?

CC @peastman

### PR DESCRIPTION
**Pull request by peastman**
Fixes #684.  This is an alternative to A3C for reinforcement learning that is claimed to be faster and require less hyperparameter tuning.  My experiences with it so far do not support those claims.  In most of my tests it is slower than A3C, not faster.  And it has all the same hyperparameters that A3C does, while also adding a couple of new ones.  Still, it's another option, and maybe other people will have more success with it than I have.

### CODE REVIEW

**Comment by rbharath** [comment]
LGTM.

@peastman Do you have any intuition for why OpenAI might find PPO to work well for them? Perhaps an artifact of their particular test cases?

**Comment by peastman** [comment]
They kind of cheat in their paper.  They don't actually compare it to A3C.  Instead they compare it to a single threaded version they call "A2C", and they compare the methods solely based on how the performance improves with the number of training steps.  They don't consider performance vs. clock time.  That hides a really important issue: PPO doesn't parallelize well at all, because it does all the optimization on one thread and leaves all the other threads idle while it's doing it.

Also, PPO runs several epochs of optimization with each set of rollouts, which means the CPU time per step is much higher.  That again makes their comparison very misleading.  They boast that PPO has much better performance than "A2C" after 1 million training steps, but don't mention that A3C could get through 4 million training steps in the same amount of clock time.

**Comment by ethancaballero** [comment]
You can bucket the rollouts like ELF does, and then the throughput will be closer to A3C:
https://github.com/facebookresearch/ELF

Look at first paragraph of Section 2 of https://arxiv.org/pdf/1707.01067.pdf 
and ctrl+f "different history length".

**Comment by peastman** [comment]
Could you explain what you mean?  That paragraph is talking about a producer-consumer model, which isn't really applicable to PPO.  In each iteration, you have to first generate all the rollouts, then repeatedly run all of them through the optimizer for multiple epochs.  And you need to complete that before you can start generating the next set of rollouts.

Maybe you just mean merging the data from multiple rollouts into a single batch?  I could try that.  It would mean fewer calls to the optimizer, which could save some CPU time, but also could mean slower learning.  And of course it would create yet another hyperparameter (the batch size) the user has to adjust.

**Comment by nikoliazekter** [comment]
DeepMind uses what they call Distributed Proximal Policy Optimization (DPPO) in their "Emergence of Locomotion Behaviours in Rich Environments". According to paper it scales well and improves over A3C in terms of wall clock time. Here it is an attempt at its implementation https://github.com/alexis-jacq/Pytorch-DPPO

**Comment by peastman** [comment]
Thanks!  I'll try doing something based on that.  They're using an older version of the PPO algorithm, so it doesn't exactly apply, and they do some things that look kind of weird to me.  But the basic idea is pretty simple.  Instead of looping over batches and doing an optimization step on each one, they have the worker threads compute the gradients for the different batches in parallel.  Then they average them, and perform a single optimization step based on the averaged gradients.  The result should be better parallel efficiency at the cost of some extra overhead.  I'll see how it works in practice.

**Comment by PaulHuangLi** [comment]
The recent paper Emergent Complexity via Multi-Agent Competition for using ppo distributedly.
PPO is improved over than A2C at a single thread ,at multithread over A3C
